flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSource.java

```diff
@@ -107,48 +127,74 @@ public Boundedness getBoundedness() {
-    public static class SourceChain<T, EnumChkT> implements Serializable {
-        final List<Tuple2<Source<T, ? extends SourceSplit, ?>, CheckpointConverter<?, ?>>> sources;
+    /** Entry for list of underlying sources. */
+    protected static class SourceListEntry implements Serializable {
+        protected final Source source;
+        protected final SourceConfigurer configurer;
 
-        public SourceChain(Source<T, ?, EnumChkT> initialSource) {
-            this(concat(Collections.emptyList(), Tuple2.of(initialSource, null)));
+        private SourceListEntry(Source source, SourceConfigurer configurer) {
+            this.source = source;
+            this.configurer = configurer;
         }
 
-        /** Construct a chain of homogeneous sources with fixed start position. */
-        public static <T, EnumChkT> SourceChain<T, EnumChkT> of(Source<T, ?, EnumChkT>... sources) {
-            Preconditions.checkArgument(sources.length >= 1, "At least one source");
-            SourceChain<T, EnumChkT> sourceChain = new SourceChain<>(sources[0]);
-            for (int i = 1; i < sources.length; i++) {
-                sourceChain = sourceChain.add(sources[i]);
-            }
-            return sourceChain;
+        public static SourceListEntry of(Source source, SourceConfigurer configurer) {
+            return new SourceListEntry(source, configurer);
         }
+    }
 
-        private SourceChain(
-                List<Tuple2<Source<T, ? extends SourceSplit, ?>, CheckpointConverter<?, ?>>>
-                        sources) {
+    /** Builder for HybridSource. */
+    public static class HybridSourceBuilder<T, EnumT extends SplitEnumerator>
+            implements Serializable {
+        private final List<SourceListEntry> sources;
+
+        public HybridSourceBuilder(Source<T, ?, ?> initialSource) {
+            this(concat(Collections.emptyList(), SourceListEntry.of(initialSource, null)));
+        }
+
+        private HybridSourceBuilder(List<SourceListEntry> sources) {
             this.sources = sources;
         }
 
-        /** Add source with fixed start position. */
-        public <NextEnumChkT> SourceChain<T, NextEnumChkT> add(
-                Source<T, ?, NextEnumChkT> nextSource) {
-            return add(nextSource, null);
+        /** Add source without switch time modification. */
+        public <ToEnumT extends SplitEnumerator, NextSourceT extends Source<T, ?, ?>>
+                HybridSourceBuilder<T, ToEnumT> addSource(NextSourceT source) {
+            return addSource(source, new NoopSourceConfigurer<>());
+        }
+
+        /** Add source with start position conversion from previous enumerator. */
+        public <ToEnumT extends SplitEnumerator, NextSourceT extends Source<T, ?, ?>>
+                HybridSourceBuilder<T, ToEnumT> addSource(
+                        NextSourceT source, SourceConfigurer<NextSourceT, EnumT> sourceConfigurer) {
+            return new HybridSourceBuilder<>(
```

Comment:
Here we could also just add to `sources` and return a cast of `this` (probably over raw). Then we also can avoid the concat and just use a mutable list inside a mutable builder.